Initial contribution of LibPressio ecosystem

robertu94 opened the pull request:
LibPressio is a generic abstraction for the compression of data.  It supports over 40 compressors and is used at over 10 institutions (that we know about).  Previously these packages were part of a 3rd party [package repository](https://github.com/robertu94/spack_packages). This PR contains packages (and modifications for 2 existing packages) to expose the LibPressio support.

I expect this PR to take some time to settle and finalize because of its size.

Modifications:

- adios2 - adds libpressio support to the existing package.
- cusz the old package was not portable and lacked key dependencies this re-write is down with the knowledge of the maintainers @dingwentao and @jtian0.

New Packages:

- arc an data corruption tool for compressed data developed by the lab of @joncalhoun40.  @joncalhoun40 would you be a maintainer for this?
- bitgroomingz a standalone implementation of bitgrooming from NetCDF packaged by @disheng222.
- fpzip a specialized compressor for floating point data developed by @lindstro and others and a predecessor to ZFP.
- libdistributed a MPI work queue with cancelation used by libpressio and libpressio_opt since this work queue is used by multiple projects outside of libpressio so is separate. Developed by @robertu94
- libpressio-adios2 a adios2 bp file reader for libpressio.  This is a separate package to solve a circular dependency.  I've worked most with @jychoi-hpc on ADIOS2
- libpressio-errorinjector allows simulating compression error with C++ standard library random distributions.  It adds additional interfaces so is separate.Developed by @robertu94
- libpressio-opt automatically configure libpressio compressors to
    meet quality objectives.   It adds additional interfaces so is
    separate.Developed by @robertu94
- libpressio-rmetric - compute libpressio metrics using the R
    programming language in a way that embeds into other programming
    languages.  It depends on a GPL dependency so is separate.
    Developed by @robertu94
- sperr and libpressio-sperr - SPERR is GPL licensed so cannot be in core
    libpresio which is under a BSD2 license. SPERR is an advanced
    compressor from the climate community.  SPERR is developed by
    @shaomeng.  @shaomeng would you like to be listed as maintainer for
    this?  LibPressio sperr is the libpressio integration.
- libpressio-tools high level tools for LibPressio like a CLI interface
    and a distributed compressor experiment frontend.  It also provides
    integration for the 3rd party compressor libraries. Developed by @robertu94
- libpressio-tthresh - tthresh is a compressor based on Higher order
    SVD. It is GPL licensed so in a separate repository. TThresh was
    written by @rballester
- libpressio the core libpressio library which supports C, C++, and
    Python and contains the bulk of the compressor plugins. Developed by
    @robertu94
- libstdcompat a backwards compatability shim for LibPressio. Developed
    by @robertu94
- mgard  an advanced compressor developed by a team at ORNL as part of
    ECP (IIRC, @qliu21 is the maintainer of MGARD, but I also interface
    with @jychoi-hpc from time to time)
- mgardx  a variant of MGARD developed by at least in part by @lxAltria
- ndzip  an advanced compressor developed by @fknorr
- nvcomp the last open source release of NVCOMP.  Unless @Nvidia starts
    providing source code again, or provides a stable public download link
    under an appropriate license we're kind of stuck at this version
- qoz a version of SZ3 that allows stable grid sizes developed by the
    team of @disheng222
- qcat a quality analysis tool for binary data developed by @disheng222
- r-libpressio R bindings for libpressio in a seperate package because
    it depends on a GPL dependency.  Developed by @robertu94
- sam2p a library that converts binary files to postscript used in
    libpressio's quality analysis example scripts.
- sol2 a library that allows embedding lua in to C++ used in LibPressio
    and LibPressioOpt to allow scripting programmatic responses
- sz-cpp a predecessor of SZ3 with a few different features supported by
    LibPressio developed by the teams of @lxAltria and @disheng222
- szauto a predecessor of SZ3 with autotuning builtin supported by
    LibPressio developed by the teams of and @disheng222 and @ayzk IIRC.
- sz3 the next generation of the SZ compressor family developed by @ayzk
    and others.
- szx an ultrafast variant of SZ developed by @ayzk, @xiaodong-yu, and
  @disheng222

robertu94 committed:
Initial contribution of LibPressio ecosystem

LibPressio is a generic abstraction for the compression of data.  It
supports over 40 compressors and is used at over 10 institutions (that
we know about).  Previously these packages were part of a 3rd party
[package repository](https://github.com/robertu94/spack_packages). This
PR contains packages (and modifications for 2 existing packages) to
expose the LibPressio support.

I expect this PR to take some time to settle and finalize because of its
size.

Modifications:

- adios2 - adds libpressio support to the existing package.
- cusz the old package was not portable and lacked key dependencies
       this re-write is down with the knowledge of the maintainers
       @dingwentao and @jtian0.

New Packages:

- arc an data corruption tool for compressed data developed by the lab
    of @joncalhoun40.  @joncalhoun40 would you be a maintainer for this?
- bitgroomingz a standalone implementation of bitgrooming from NetCDF
    packaged by @disheng222.
- fpzip a specialized compressor for floating point data developed by
    @lindstro and others and a predecessor to ZFP.
- libdistributed a MPI work queue with cancelation used by libpressio
    and libpressio_opt since this work queue is used by multiple projects
    outside of libpressio so is separate. Developed by @robertu94
- libpressio-adios2 a adios2 bp file reader for libpressio.  This is a
    separate package to solve a circular dependency.  I've worked most
    with @jychoi-hpc on ADIOS2
- libpressio-errorinjector allows simulating compression error with C++
    standard library random distributions.  It adds additional
    interfaces so is separate.Developed by @robertu94
- libpressio-opt automatically configure libpressio compressors to
    meet quality objectives.   It adds additional interfaces so is
    separate.Developed by @robertu94
- libpressio-rmetric - compute libpressio metrics using the R
    programming language in a way that embeds into other programming
    languages.  It depends on a GPL dependency so is separate.
    Developed by @robertu94
- sperr and libpressio-sperr - SPERR is GPL licensed so cannot be in core
    libpresio which is under a BSD2 license. SPERR is an advanced
    compressor from the climate community.  SPERR is developed by
    @shaomeng.  @shaomeng would you like to be listed as maintainer for
    this?  LibPressio sperr is the libpressio integration.
- libpressio-tools high level tools for LibPressio like a CLI interface
    and a distributed compressor experiment frontend.  It also provides
    integration for the 3rd party compressor libraries. Developed by @robertu94
- libpressio-tthresh - tthresh is a compressor based on Higher order
    SVD. It is GPL licensed so in a separate repository. TThresh was
    written by @rballester
- libpressio the core libpressio library which supports C, C++, and
    Python and contains the bulk of the compressor plugins. Developed by
    @robertu94
- libstdcompat a backwards compatability shim for LibPressio. Developed
    by @robertu94
- mgard  an advanced compressor developed by a team at ORNL as part of
    ECP (IIRC, @qliu21 is the maintainer of MGARD, but I also interface
    with @jychoi-hpc from time to time)
- mgardx  a variant of MGARD developed by at least in part by @lxAltria
- ndzip  an advanced compressor developed by @fknorr
- nvcomp the last open source release of NVCOMP.  Unless @Nvidia starts
    providing source code again, or provides a stable public download link
    under an appropriate license we're kind of stuck at this version
- qoz a version of SZ3 that allows stable grid sizes developed by the
    team of @disheng222
- qcat a quality analysis tool for binary data developed by @disheng222
- r-libpressio R bindings for libpressio in a seperate package because
    it depends on a GPL dependency.  Developed by @robertu94
- sam2p a library that converts binary files to postscript used in
    libpressio's quality analysis example scripts.
- sol2 a library that allows embedding lua in to C++ used in LibPressio
    and LibPressioOpt to allow scripting programmatic responses
- sz-cpp a predecessor of SZ3 with a few different features supported by
    LibPressio developed by the teams of @lxAltria and @disheng222
- szauto a predecessor of SZ3 with autotuning builtin supported by
    LibPressio developed by the teams of and @disheng222 and @ayzk IIRC.
- sz3 the next generation of the SZ compressor family developed by @ayzk
    and others.
- szx an ultrafast variant of SZ developed by @ayzk, @xiaodong-yu, and
  @disheng222

robertu94 committed:
Initial contribution of LibPressio ecosystem

LibPressio is a generic abstraction for the compression of data.  It
supports over 40 compressors and is used at over 10 institutions (that
we know about).  Previously these packages were part of a 3rd party
[package repository](https://github.com/robertu94/spack_packages). This
PR contains packages (and modifications for 2 existing packages) to
expose the LibPressio support.

I expect this PR to take some time to settle and finalize because of its
size.

Modifications:

- adios2 - adds libpressio support to the existing package.
- cusz the old package was not portable and lacked key dependencies
       this re-write is down with the knowledge of the maintainers
       @dingwentao and @jtian0.

New Packages:

- arc an data corruption tool for compressed data developed by the lab
    of @joncalhoun40.  @joncalhoun40 would you be a maintainer for this?
- bitgroomingz a standalone implementation of bitgrooming from NetCDF
    packaged by @disheng222.
- fpzip a specialized compressor for floating point data developed by
    @lindstro and others and a predecessor to ZFP.
- libdistributed a MPI work queue with cancelation used by libpressio
    and libpressio_opt since this work queue is used by multiple projects
    outside of libpressio so is separate. Developed by @robertu94
- libpressio-adios2 a adios2 bp file reader for libpressio.  This is a
    separate package to solve a circular dependency.  I've worked most
    with @jychoi-hpc on ADIOS2
- libpressio-errorinjector allows simulating compression error with C++
    standard library random distributions.  It adds additional
    interfaces so is separate.Developed by @robertu94
- libpressio-opt automatically configure libpressio compressors to
    meet quality objectives.   It adds additional interfaces so is
    separate.Developed by @robertu94
- libpressio-rmetric - compute libpressio metrics using the R
    programming language in a way that embeds into other programming
    languages.  It depends on a GPL dependency so is separate.
    Developed by @robertu94
- sperr and libpressio-sperr - SPERR is GPL licensed so cannot be in core
    libpresio which is under a BSD2 license. SPERR is an advanced
    compressor from the climate community.  SPERR is developed by
    @shaomeng.  @shaomeng would you like to be listed as maintainer for
    this?  LibPressio sperr is the libpressio integration.
- libpressio-tools high level tools for LibPressio like a CLI interface
    and a distributed compressor experiment frontend.  It also provides
    integration for the 3rd party compressor libraries. Developed by @robertu94
- libpressio-tthresh - tthresh is a compressor based on Higher order
    SVD. It is GPL licensed so in a separate repository. TThresh was
    written by @rballester
- libpressio the core libpressio library which supports C, C++, and
    Python and contains the bulk of the compressor plugins. Developed by
    @robertu94
- libstdcompat a backwards compatability shim for LibPressio. Developed
    by @robertu94
- mgard  an advanced compressor developed by a team at ORNL as part of
    ECP (IIRC, @qliu21 is the maintainer of MGARD, but I also interface
    with @jychoi-hpc from time to time)
- mgardx  a variant of MGARD developed by at least in part by @lxAltria
- ndzip  an advanced compressor developed by @fknorr
- nvcomp the last open source release of NVCOMP.  Unless @Nvidia starts
    providing source code again, or provides a stable public download link
    under an appropriate license we're kind of stuck at this version
- qoz a version of SZ3 that allows stable grid sizes developed by the
    team of @disheng222
- qcat a quality analysis tool for binary data developed by @disheng222
- r-libpressio R bindings for libpressio in a seperate package because
    it depends on a GPL dependency.  Developed by @robertu94
- sam2p a library that converts binary files to postscript used in
    libpressio's quality analysis example scripts.
- sol2 a library that allows embedding lua in to C++ used in LibPressio
    and LibPressioOpt to allow scripting programmatic responses
- sz-cpp a predecessor of SZ3 with a few different features supported by
    LibPressio developed by the teams of @lxAltria and @disheng222
- szauto a predecessor of SZ3 with autotuning builtin supported by
    LibPressio developed by the teams of and @disheng222 and @ayzk IIRC.
- sz3 the next generation of the SZ compressor family developed by @ayzk
    and others.
- szx an ultrafast variant of SZ developed by @ayzk, @xiaodong-yu, and
  @disheng222

robertu94 commented:
@scheibelp You asked me to ping you when I submitted this.

robertu94 commented:
@spackbot style fix

robertu94 commented:
@spackbot fix style

robertu94 committed:
fix many package audits and add maintainers for spack packages

robertu94 committed:
try to fix package license headers

robertu94 committed:
missed one

shaomeng commented on var/spack/repos/builtin/packages/sperr/package.py:
```diff
@@ -0,0 +1,39 @@
+    """SPERR is a lossy scientific (floating-point) data compressor that can
+    perform either error-bounded or size-bounded data compression"""
+
+    homepage = "https://github.com/shaomeng/SPERR"
```
```suggestion
    homepage = "https://github.com/NCAR/SPERR"
```

shaomeng commented on var/spack/repos/builtin/packages/sperr/package.py:
```diff
@@ -0,0 +1,39 @@
+    depends_on("pkgconf", type=("build"), when="+zstd")
+
+    variant("shared", description="build shared libaries", default=True)
+    variant("qz", description="coding terminates by quantization level", default=True)
```
This variant is no longer there... We can remove it now I believe.

shaomeng commented on var/spack/repos/builtin/packages/sperr/package.py:
```diff
@@ -0,0 +1,39 @@
+    def cmake_args(self):
+        args = [
+            self.define_from_variant("BUILD_SHARED_LIBS", "shared"),
+            self.define_from_variant("QZ_TERM", "qz"),
```
Similarly, this option can be removed. 

However, this is an option `USE_OMP` which if not specified, will be turned on by default. I'm not sure if you want to include that one or not.

robertu94 replied:
What would you want the default to be? I would assume on?

shaomeng replied:
The CMake option, `QZ_TERM`, is no longer there, so we should completely delete it. 

I was suggesting that another CMake option in SPERR, `USE_OMP`, should be set `ON` if possible. ;-) (I'm not quite sure what syntax should be used here...)

shaomeng reviewed (commented):
Hi @robertu94 , thanks for including SPERR in the newest LibPressio and I'm happy to be listed as a maintainer for SPERR. Also, I've suggested a few changes, largely because the software has evolved since we spoke last time. Please take a look and let me know what you think!

robertu94 committed:
Update var/spack/repos/builtin/packages/sperr/package.py

Co-authored-by: Samuel Li <shaomeng@users.noreply.github.com>

shaomeng commented on var/spack/repos/builtin/packages/sperr/package.py:
```diff
@@ -0,0 +1,39 @@
+    depends_on("pkgconf", type=("build"), when="+zstd")
+
+    variant("shared", description="build shared libaries", default=True)
+    variant("qz", description="coding terminates by quantization level", default=True)
```
```suggestion
```

robertu94 committed:
suggested fixes to sperr

robertu94 committed:
add  shaomeng  as a maintainer for SPERR

scheibelp commented on var/spack/repos/builtin/packages/cusz/package.py:
```diff
@@ -6,19 +6,22 @@
+    conflicts("~cuda")
+    conflicts("cuda_arch=none", when="+cuda")
+
+    version("develop", branch="develop")
```
Generally we discourage removing older versions: this can cause trouble if a user has already installed the older package. I don't see any packages which depend on it though, so this is less of an issue.

Also, if there is any recent fixed version (e.g. with a tag), that would be ideal, since `@develop` isn't reproducible (i.e. you are pulling the tip of the branch).

robertu94 replied:
In the past, I've had issues getting the older versions to build consistently.  I suspect that people aren't using it widely yet.

>  any recent fixed version ... would be ideal since `@develop` isn't reproducible.

@jtian0 or @dingwentao  Could you please make a release of cuSZ containing the fixes I sent this spring so I can add it to this PR.

robertu94 replied:
done for now.  We'll send the HIP stuff in a later PR

scheibelp commented on var/spack/repos/builtin/packages/digitrounding/package.py:
```diff
@@ -0,0 +1,30 @@
+    maintainers = ["robertu94"]
+
+    version("master", branch="master")
+    version("2020-27-20", commit="7b18679aded7a85e6f221f7f5cd4f080f322bc33")
```
Sanity check: is this supposed to by a YYYY-MM-DD? If so, it looks like there was a typo

robertu94 replied:
Yes.  Good catch. Will be fixed in my next push

scheibelp commented on var/spack/repos/builtin/packages/libpressio-frsz/package.py:
```diff
@@ -0,0 +1,24 @@
+    """Fized Rate SZ"""
+
+    homepage = "https://github.com/robertu94/frsz"
+    url = "https://github.com/robertu94/frsz"
```
This link doesn't resolve for me: I don't see any repo in https://github.com/robertu94?tab=repositories called `frsz`. I do see an `sz` repo.

robertu94 replied:
Sorry.  This is a pre-release package, and it shouldn't have been included in this PR. I'll remove it for now.

scheibelp commented on var/spack/repos/builtin/packages/libpressio-rmetric/package.py:
```diff
@@ -0,0 +1,41 @@
+    depends_on("r-rinside")
+
+    def cmake_args(self):
+        args = ["HAS_SPACK=ON"]
```
(question) out of curiosity, how does the CMake behavior differ for `HAS_SPACK=ON`?

robertu94 replied:
It changes the way that R is found.  IIRC, on some systems that we used, R_HOME was set to some bogus value which results in spack not finding R or finding the wrong R (I can't remember which).  If we can find a fix (and a machine where I can reproduce this), I'd be happy to remove this detection.

tgamblin replied:
Can you change things so that the R package sets `R_HOME` properly, to avoid having a `HAS_SPACK` special case in your build?

robertu94 replied:
Done

scheibelp commented on var/spack/repos/builtin/packages/libpressio-sperr/package.py:
```diff
@@ -0,0 +1,24 @@
+
+    homepage = "https://github.com/robertu94"
+    url = "https://github.com/robertu94/libpressio-sperr/archive/refs/tags/0.0.1.tar.gz"
+    git = url
```
I don't think Spack can treat the `url` (which specifies an archive) as a git url, it should be `https://github.com/robertu94/libpressio-sperr.git`.

robertu94 replied:
Yep. fixed.

scheibelp commented on var/spack/repos/builtin/packages/libpressio-tools/package.py:
```diff
@@ -0,0 +1,96 @@
+
+    depends_on("mpi", when="+mpi")
+    depends_on("libpressio+hdf5+libdistributed+mpi", when="+mpi")
+    depends_on("libpressio+hdf5", when="~mpi")
```
The prior two statements could be expressed as 

```
depends_on("libpressio+hdf5")
depends_on("libpressio+libdistributed+mpi", when="+mpi")
```

robertu94 replied:
fixed

scheibelp commented on var/spack/repos/builtin/packages/mgard/package.py:
```diff
@@ -0,0 +1,41 @@
+
+    version("master", branch="master")
+    version("robertu94", git="https://github.com/robertu94/MGARD", branch="master", prefered=True)
+    version("2020-10-01", commit="b67a0ac")
```
Can you use the full commit hash here?

robertu94 replied:
done

scheibelp commented on var/spack/repos/builtin/packages/mgard/package.py:
```diff
@@ -0,0 +1,41 @@
+    maintainers = ["robertu94"]
+
+    version("master", branch="master")
+    version("robertu94", git="https://github.com/robertu94/MGARD", branch="master", prefered=True)
```
Do you anticipate this being generally useful, or would you be able to maintain an upstream package repo that defines this additional version?

I have a similar question for analogous cases in this PR like `nvcomp`.

robertu94 replied:
One of the challenges with MGARD is that their developers evolve their API ways that unintentionally break backwards compatibility or occasionally undoing fixes that I do to get things to work in spack.  I understand it's research code, but I was fielding user support requests frequently for LibPressio because MGARD broke compatibility so often so I forked it to make it stable.

With `nvcomp`, NVIDIA took things closed source and broke several packages that were using it (by removing headers among other things), and as far as I can tell never included my fixes I submitted to them to support spack in an open source release (it was in the closed source version).  My fork fixes their cmake problems and uses the last open source version.  Since I have no control over Nvidia packages this and they break stuff in this package frequently (either urls, package ids, etc...).  If someone from @NVIDIA can do a better job and wants to provide the code under an open source license again, I welcome their contribution.

scheibelp commented on var/spack/repos/builtin/packages/r-libpressio/package.py:
```diff
@@ -0,0 +1,35 @@
+    depends_on("r@3.3.0:", type=("build", "run"))
+    depends_on("r-rcpp", type=("build", "link", "run"))
+    depends_on("libpressio+json", type=("build", "link", "run"))
+    depends_on("libpressio@0.65.0:+json", type=("build", "link", "run"), when="@1.2:")
```
(minor) Since you have

```
depends_on("libpressio+json", type=("build", "link", "run"))
```

one line above, you can remove `+json` from this line, like:

```
depends_on("libpressio@0.65.0:", type=("build", "link", "run"), when="@1.2:")
```

robertu94 replied:
fixed

scheibelp commented on var/spack/repos/builtin/packages/r-libpressio/package.py:
```diff
@@ -0,0 +1,35 @@
+    homepage = "https://github.com/robertu94/libpressio-r"
+    url = "https://github.com/robertu94/libpressio-r/archive/0.0.1.tar.gz"
+
+    # notify when the package is updated.
```
(minor) I think this comment should be removed.

robertu94 replied:
fixed

scheibelp commented on var/spack/repos/builtin/packages/sam2p/package.py:
```diff
@@ -0,0 +1,24 @@
+
+
+class Sam2p(Package):
+    """A raster to PostScript/PDF image conversn program"""
```
(typo) conversn -> conversion

robertu94 replied:
fixed

scheibelp commented on var/spack/repos/builtin/packages/sol2/package.py:
```diff
@@ -0,0 +1,22 @@
+    """a C++ <-> Lua API wrapper with advanced features and top notch performance"""
+
+    homepage = "https://github.com/ThePhD/sol2"
+    url = "https://github.com/ThePhD/sol2/archive/v3.0.3.tar.gz"
```
There is already a `lua-sol2` package.

FWIW in my experience the easiest way to know if a Spack package has been made yet is to look for the "identifying" part of the URL (In this case I recursively grep'ed through our builtin package repo for "sol2")

robertu94 replied:
I looked with spack list, and I didn't see it.  I could have been added after I first packaged LibPressio-Opt/OptZConfig since that was one of the earlier packages I added.

robertu94 replied:
I've now switched to the upstream lua-sol2

scheibelp commented on var/spack/repos/builtin/packages/sperr/package.py:
```diff
@@ -0,0 +1,43 @@
+    perform either error-bounded or size-bounded data compression"""
+
+    homepage = "https://github.com/NCAR/SPERR"
+    git = "https://github.com/robertu94/SPERR"
```
If your git repo is used as a staging area for changes that you eventually intend to upstream to https://github.com/NCAR/SPERR, then I think this should refer to that (elsewhere I suggest that you can create an inherited package if you have bugfixes that you need to use in the meantime - I'm curious if that would work here as well).

I'm curious if other packages following this theme (e.g. `sz-cpp`, `qoz`, `libpressio`, and `mgardx`, the last of which is actually a fork of a fork) can do this as well.

robertu94 replied:
Working with @shaomeng has been good and he's pretty quickly adopted fixes for spack.  We can probably switch over to his upstream.

`sz-cpp` became SZ3 but is different enough that some people want to compare against it separately.  They have different APIs so can't really be treated as the same.

`qoz` should eventually get merged into SZ3, but I am not sure of the timeline on this.

`libpressio` the development version is `robertu94` and the stable version is `codarcode`  I update the latter about 1/year.  `robertu94` gets small updates ~monthly.

`mgardx` was a research code based on MGARD with a different API.  Could someone (maybe  @qliu21) confirm if it will be merged into MGARD?  For now its a seperate package.

JieyangChen7 replied:
@robertu94 I'm the main developer of `mgardx`. `mgardx` and the original `mgard` are in the same repository but they have two sets of APIs. `mgard` is the CPU-only version and `mgardx` is the CPU-GPU portable version, which make their internal design significantly different. In the foreseeable future,  we have no plan to merge them together.  For the propose of optimizing for Exascale computing, we are using `mgardx` as the default version and its API is used to integrate with ADIOS2.

robertu94 replied:
Has there been a release since this has happened? I can update LibPressio accordingly if it has

shaomeng replied:
For SPERR, I'm planning to put out a release in 2-3 weeks. Maybe I'll ping you when I do so you can update LibPressio?

robertu94 replied:
@shaomeng Sure. Are there api changes, or just internal changes?

shaomeng replied:
> internal c

There will not be API change, just internal tweaking, for the best I can tell. I can sync up with you again to run tests locally to make sure libpressio works well.

robertu94 replied:
@scheibelp I converted as many packages as practical to use upstream sources, and I pinned the versions.  Any package with a `libpressio-*` name is a package for which I am the main developer so packages with this prefix should have a `robertu94` git URL.  There are a few more remaining for which using upstream repos are not practical for reasons we discussed on the Spack call.  If the developers of these packages want to take over maintenance and provide stability under an open source licence, I'm happy to re-direct these remaining packages to point to their branches in the future.

scheibelp commented on var/spack/repos/builtin/packages/szx/package.py:
```diff
@@ -0,0 +1,27 @@
+class Szx(AutotoolsPackage):
+    """An ultra fast error bounded compressor for scientific datasets"""
+
+    homepage = "https://github.com/szcompressor/szx"
```
I don't see an `szx` sub repository at "https://github.com/szcompressor/szx"

robertu94 replied:
@disheng222 when will szx be released?  I can make this private if it going to be a while.

xiaodong-yu replied:
@robertu94 This work was partially supported by the fund from Aramco. We need to ask for their permission for release. Will keep you posted.

scheibelp commented on var/spack/repos/builtin/packages/szx/package.py:
```diff
@@ -0,0 +1,27 @@
+
+    homepage = "https://github.com/szcompressor/szx"
+    url = "https://github.com/szcompressor/szx"
+    git = "ssh://git@github.com/szcompressor/szx"
```
Note you will generally be prompted for password etc. unless you have set up passwordless ssh authentication with github (FYI if you have, other users would likely be prompted for credentials): I recommend an `https`-based URL

robertu94 replied:
This package was private when this package was created.  I'll update the URL if this becomes public. (see above)

scheibelp replied:
To be clear: I'm OK waiting for that, but I wouldn't want to merge the PR until either:

* The repo becomes public (and the `git` property is changed) or...
* This package is removed

It would be problematic to add packages which other users cannot use.

robertu94 replied:
@scheibelp of course. Based on @xiaodong-yu response above, I'll push a commit that removes this until we we get approval for release.

robertu94 replied:
should be gone now.

scheibelp commented on var/spack/repos/builtin/packages/libpressio-errorinjector/package.py:
```diff
@@ -0,0 +1,23 @@
+
+
+class LibpressioErrorinjector(CMakePackage):
+    """LibPressioErrorInjector injects errors in to data for sensitivity studies"""
```
(typo) "in to" -> "into"

robertu94 replied:
fixed

scheibelp reviewed (changes requested):
High level notes:

* There is at least one package here (`sol2`) which already has a Spack package. I checked several other packages but not all of them in this PR.
* Several packages retrieve source from your personal repository and I'm curious if you can pull from the repo they fork from (if you are in fact the primary maintainer of these repos, then using your forks makes sense)
* A couple of links appear to be broken (I mention this in specific comments)

robertu94 committed:
A round of fixes based on feedback from @scheibelp

robertu94 commented:
@scheibelp I checked the other packages in the PR, and `sol2` was the only one that appears to be a duplicate now.

robertu94 committed:
remove szx for now

robertu94 committed:
Various update to libpressio ecosystem

+ new package versions
+ found a dependency `dlib` that didn't correspond to the same `dlib`
  that libpressio-opt refers to.  Renamed the old package because it had
  way fewer users, and davisking/dlib hold the dlib name in Ubuntu,
  Fedora, Homebrew, Pip.  Discused with @adamjstewart who added the old
  dlib package, and he was okay with the replacement
+ added missing fix for CUB relating to #32547 needed for cuSZ to build
  using Spack.

robertu94 committed:
fix libpressio-r metric and isolate R_HOME

robertu94 committed:
nvcomp fixes and support

robertu94 committed:
various package fixes for libpressio

+ Removed hack for R_HOME/R_ENVION in lipbressio-rmetric
+ Pin versions rather than rely on branches
+ Use upstream repositories whenever possible

robertu94 commented:
@scheibelp I think I have completed all requested changes.  Can you please take another look?  Also the CI failures seems to be github related and not related to these changes specifically:

> diff is temporarily unavailable due to heavy server load

scheibelp commented on var/spack/repos/builtin/packages/libpressio-sperr/package.py:
```diff
@@ -0,0 +1,25 @@
+
+    homepage = "https://github.com/robertu94/libpressio-sperr"
+    url = "https://github.com/robertu94/libpressio-sperr/archive/refs/tags/0.0.1.tar.gz"
+    git = homepage
```
(the git url is broken)

robertu94 replied:
Also works for me?

```console
[runderwood@defiant tmp]$ git clone "https://github.com/robertu94/libpressio-sperr"
Cloning into 'libpressio-sperr'...
remote: Enumerating objects: 22, done.
remote: Counting objects: 100% (22/22), done.
remote: Compressing objects: 100% (12/12), done.
remote: Total 22 (delta 6), reused 20 (delta 4), pack-reused 0
Receiving objects: 100% (22/22), 16.63 KiB | 1.28 MiB/s, done.
Resolving deltas: 100% (6/6), done.
[runderwood@defiant tmp]$ curl -LO "https://github.com/robertu94/libpressio-sperr/archive/refs/tags/0.0.1.tar.gz"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 15186    0 15186    0     0  52876      0 --:--:-- --:--:-- --:--:-- 52876

```

scheibelp replied:
My mistake: I misread this as setting `git = url` rather than `git = homepage`

scheibelp commented on var/spack/repos/builtin/packages/libpressio-tthresh/package.py:
```diff
@@ -0,0 +1,39 @@
+
+    homepage = "https://github.com/robertu94/libpressio_tthresh"
+    url = "https://github.com/robertu94/libpressio_tthresh/archive/refs/tags/0.0.1.tar.gz"
+    git = homepage
```
(broken git url)

robertu94 replied:
Works for me?  Was this meant for another repo?

```
[runderwood@defiant tmp]$ git clone https://github.com/robertu94/libpressio_tthresh
Cloning into 'libpressio_tthresh'...
remote: Enumerating objects: 63, done.
remote: Counting objects: 100% (63/63), done.
remote: Compressing objects: 100% (42/42), done.
remote: Total 63 (delta 22), reused 59 (delta 18), pack-reused 0
Receiving objects: 100% (63/63), 1.45 MiB | 19.75 MiB/s, done.
Resolving deltas: 100% (22/22), done.
[runderwood@defiant tmp]$ curl -LO https://github.com/robertu94/libpressio_tthresh/archive/refs/tags/0.0.1.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 1490k    0 1490k    0     0  3398k      0 --:--:-- --:--:-- --:--:-- 3398k
```

scheibelp commented on var/spack/repos/builtin/packages/mgard/package.py:
```diff
@@ -0,0 +1,41 @@
+    """MGARD error bounded lossy compressor"""
+
+    homepage = "https://github.com/CODARcode/MGARD"
+    git = "https://github.com/robertu94/MGARD"
```
In cases where you are maintaining a fork, can you mention the original repository in the docstring?

A short explanation of why the fork is used would be helpful (e.g. in the case of `nvcomp` that they went closed-source).

robertu94 replied:
@scheibelp please take a look at the most recent commit.

scheibelp reviewed (changes requested):
~There's a couple git references that are broken.~ Also, it would be great if you could clarify in packages where you maintain a personal fork that you are doing so in the docstring (also if feasible with an explanation of why).

robertu94 committed:
clarified when and where forks were used

robertu94 commented:
@scheibelp I looked at the CI failures.  I looked at many of them, and all happened in Adios2.  As far as I can tell these are unrelated to the changes we made in Adios2.  The exact error message is:

```console
/tmp/root/spack-stage/spack-stage-adios2-2.7.1-tbawoed3qm3m5xqq5exm6fghdnkmam3d/spack-src/source/adios2/helper/adiosMemory.cpp:32:22: error: no match for 'operator==' (operand types are 'const adios2::DataType' and 'const char [1]')
         if (destType == "")
             ~~~~~~~~~^~~~~
make[2]: *** [source/adios2/CMakeFiles/adios2_core.dir/helper/adiosMemory.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....```

scheibelp commented:
@robertu94 Could you try setting `default=False` for the `+libpressio` variant in `adios2`: that would confirm whether there might be some difference in the CI environment which prevents that from building successfully.

robertu94 committed:
try adios without libpressio

scheibelp commented:
@robertu94 if you are able to try reverting the latest commit, we can see if this was a temporary error, but I suspect the changes to `adios2` may in fact be an issue. We can also merge this as is (leaving `~libpressio` as the default) if you are able to build with `+libpressio` yourself (and you can update `adios2` in a later PR to make `+libpressio` the default).

robertu94 commented:
@scheibelp It builds for me.  Let's go ahead and merge this now (with `adios2~libpressio`) and handle the latter in a follow up later.

```console
[runderwood@defiant ~]$ spack install adios2+libpressio ^ sz3
[+] /usr (external diffutils-3.8-cmrhmnpvzetjh4wqaveie64jxqq2ssfk)
[+] /usr (external cmake-3.22.2-yoahtgzvn7dkcx5324hjg5zuggvtlcjr)
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/lz4-1.9.3-t5j62ojcexh336hgz5qm4443s3glm5ps
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/zlib-1.2.12-itsi2i4tzpvkirehmpm4jjffkhtq6ooc
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/zstd-1.5.2-htznodvlcs5qakvtbnkl67ceg7q27iql
[+] /usr (external bison-3.8.2-7jz6xa76ansjnduissm2ent3z44oio3l)
[+] /usr (external flex-2.6.4-zff3q4r4itu2heylabncthhsheisht4s)
[+] /usr (external m4-1.4.19-sjyxxeuj7m4chyhjjkjuncqsypascr5x)
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/pkg-config-0.29.2-2sqlxuvivzcplfjzerwkcjislinhcxan
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/libmd-1.0.4-kw2q33lp5noau3xh2x2hsn2qnt6sr2yp
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/libiconv-1.16-opvjp3hx2hutzswdnem32battgxgoc2b
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/xz-5.2.5-q7fw4xs7gr7pkxtfaecdvvwl32wxccgl
[+] /usr (external tar-1.34-imgd4ijz4m7gmk55rb6afiu7huysfcty)
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/libffi-3.4.2-kq3zs7rl7udxhaszrzvdvw7feuadb5m3
[+] /usr (external openssl-3.0.5-va446ch5fjo4qlggbp6iclpgdwhlywat)
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/gsl-2.7.1-6qxuvfmjcopnkiwiqhvfbmngfizuzjy4
[+] /usr (external findutils-4.9.0-loprc5ure6udqaj556zfw5hrva2yieff)
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/util-macros-1.19.3-brffbhnadauippxxk5pdwajmxkdjhtbs
[+] /usr (external autoconf-2.71-5nareulkxpgffxqi26eyjkzs2qjwp4yq)
[+] /usr (external automake-1.16.5-4dacek4lg433rt333kehaasumzuvthgf)
[+] /usr (external libtool-2.4.7-xlut7mvuolw7wpsa3tossu2u2vmhm6ip)
[+] /usr (external m4-1.4.19-4iyjtisldmzbmwgc6rh64r4u5csb5cvi)
[+] /usr (external tar-1.34-3433k5ltb2nzx75k6ssedprvrlwo76i5)
[+] /usr (external openssl-3.0.5-mkzafu5op32esconrswcoouhyd2bxysr)
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/bzip2-1.0.8-mf6yxiy7vp5in2szmobwn2556d4puxn6
==> Installing zfp-0.5.5-qvjmkfzezpkdlyp3ytl4zlhlfmhhtr4e
==> No binary for zfp-0.5.5-qvjmkfzezpkdlyp3ytl4zlhlfmhhtr4e found: installing from source
==> Using cached archive: /home/runderwood/git/contrib/spack/var/spack/cache/_source-cache/archive/fd/fdf7b948bab1f4e5dccfe2c2048fd98c24e417ad8fb8a51ed3463d04147393c5.tar.gz
==> No patches needed for zfp
==> zfp: Executing phase: 'cmake'
==> zfp: Executing phase: 'build'
==> zfp: Executing phase: 'install'
==> zfp: Successfully installed zfp-0.5.5-qvjmkfzezpkdlyp3ytl4zlhlfmhhtr4e
  Fetch: 0.00s.  Build: 3.62s.  Total: 3.62s.
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/zfp-0.5.5-qvjmkfzezpkdlyp3ytl4zlhlfmhhtr4e
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/snappy-1.1.9-k2x4226qpxyplaklpgciawqfym7dfdaj
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/libstdcompat-0.0.15-vnqxcovbimynryngnwgurtihy5zdiy7l
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/libpng-1.6.37-ei2satrjdtmqht25aocudmpouzawd6l6
==> Installing sz-2.1.12.2-3tpk6ojpefei73i5ekdmrtn3iagrlq46
==> No binary for sz-2.1.12.2-3tpk6ojpefei73i5ekdmrtn3iagrlq46 found: installing from source
==> Using cached archive: /home/runderwood/git/contrib/spack/var/spack/cache/_source-cache/archive/42/427e263e1fed1b0a56e13e0aff8e6a19c6d78d5f35dd16856876c70ab6066dc6.tar.gz
==> No patches needed for sz
==> sz: Executing phase: 'cmake'
==> sz: Executing phase: 'build'
==> sz: Executing phase: 'install'
==> sz: Successfully installed sz-2.1.12.2-3tpk6ojpefei73i5ekdmrtn3iagrlq46
  Fetch: 0.01s.  Build: 13.08s.  Total: 13.09s.
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/sz-2.1.12.2-3tpk6ojpefei73i5ekdmrtn3iagrlq46
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/libnl-3.3.0-um6spgfa2wfugtht4uo2wgk77vpcuykh
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/libtool-2.4.7-wx7hqjbqkmrj6y4742dlvwke3imm4opw
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/ncurses-6.2-4mzcgqljnvgmwgyvphshs2qbgrbjblsr
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/util-linux-uuid-2.37.4-i77ukptbd4nk3lhw7cdpz6jbsp3vncz5
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/libbsd-0.11.5-ocorunpo4wvtutnru73igc234rh2zpl4
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/libxml2-2.9.13-2ecdaroxqcwmvnyr36jotfpcqopwmbfg
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/sz3-3.1.5.4-axrs3ixd2etbf7ercl33nclqhhjygdc6
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/c-blosc-1.21.1-2tmwwb23neosp5xz25wvg7c7dusifjom
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/libpciaccess-0.16-oullxthmgmmffibvvnehffrjbtzlsrt5
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/readline-8.1.2-2xekztcft4i3qhsjcawjl3hqfdjncoyl
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/expat-2.4.8-qkousqa4bhqqaj6k72d3rr2r3xfqnfjy
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/gettext-0.21-gopkzmfd6vgvxzqqp535utf5pozw65id
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/gettext-0.21-5i4ihvd2hwwirh54wwqis5hcuaztj4oh
==> Installing libpressio-0.88.0-oujutp6554ghll6moqyz7ljv5msuzntd
==> No binary for libpressio-0.88.0-oujutp6554ghll6moqyz7ljv5msuzntd found: installing from source
==> Using cached archive: /home/runderwood/git/contrib/spack/var/spack/cache/_source-cache/archive/43/4358441f0d10559d571327162a216617d16d09569a80e13ad286e3b7c41c5b9b.tar.gz
==> No patches needed for libpressio
==> libpressio: Executing phase: 'cmake'
==> libpressio: Executing phase: 'build'
==> libpressio: Executing phase: 'install'
==> libpressio: Successfully installed libpressio-0.88.0-oujutp6554ghll6moqyz7ljv5msuzntd
  Fetch: 0.00s.  Build: 5m 9.72s.  Total: 5m 9.72s.
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/libpressio-0.88.0-oujutp6554ghll6moqyz7ljv5msuzntd
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/hwloc-2.7.1-qlcsa55rtgouzrtwv5ttgdw2eu5geuth
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/sqlite-3.38.5-apalj6gzw3ei6ffcx3qdj5rroupcq64x
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/gdbm-1.19-opkw2gqyl3gejgw6h7vurkyvohxjzegc
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/python-3.9.13-pxf7n5fzzra6kizlwlt7urk6gbqichca
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/python-3.8.13-4cxgwmduyr33xd4uvr624f32oty6weg2
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/yaksa-0.2-rfx2vxmknpa5eu2zjmvgfzxdzt5epbez
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/py-pip-21.3.1-3uf2g56rytmvs46hwgflgvffrkzuafoc
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/py-setuptools-63.0.0-utuqdev43aiecsz6oj7op6h5m3awldxl
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/py-wheel-0.37.0-gaenfizagxzmrzao5db2s4zcfcjgqhwc
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/py-docutils-0.18.1-v2oasbc66aecps4t4el2fu33bemxkvbb
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/rdma-core-39.1-shwtocpyncyz3z46zlv7vo2w4w7jtgux
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/libfabric-1.14.1-7fxswahuf6pofnbbwjywamsz5aqbk27x
==> Installing mpich-4.0.2-w2omnlxraxv5jtxmmvyt7kd64j23vhzc
==> No binary for mpich-4.0.2-w2omnlxraxv5jtxmmvyt7kd64j23vhzc found: installing from source
==> Using cached archive: /home/runderwood/git/contrib/spack/var/spack/cache/_source-cache/archive/5a/5a42f1a889d4a2d996c26e48cbf9c595cbf4316c6814f7c181e3320d21dedd42.tar.gz
==> Using cached archive: /home/runderwood/git/contrib/spack/var/spack/cache/_source-cache/archive/d4/d4c0e99a80f6cb0cb0ced91f6ad5da776c4a70f70f805f08096939ec9a92483e
==> Applied patch https://github.com/pmodels/mpich/commit/3a1f618e017547c9710ab4fb01ae258a01477190.patch?full_index=1
==> mpich: Executing phase: 'autoreconf'
==> mpich: Executing phase: 'configure'
==> mpich: Executing phase: 'build'
==> mpich: Executing phase: 'install'
==> mpich: Successfully installed mpich-4.0.2-w2omnlxraxv5jtxmmvyt7kd64j23vhzc
  Fetch: 0.06s.  Build: 7m 20.49s.  Total: 7m 20.56s.
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/mpich-4.0.2-w2omnlxraxv5jtxmmvyt7kd64j23vhzc
==> Installing adios2-2.8.3-m7datpc2p3xb3ilgnmst2tpgiz4rf32w
==> No binary for adios2-2.8.3-m7datpc2p3xb3ilgnmst2tpgiz4rf32w found: installing from source
==> Fetching https://github.com/ornladios/ADIOS2/archive/v2.8.3.tar.gz
==> No patches needed for adios2
==> adios2: Executing phase: 'cmake'
==> adios2: Executing phase: 'build'
==> adios2: Executing phase: 'install'
==> adios2: Successfully installed adios2-2.8.3-m7datpc2p3xb3ilgnmst2tpgiz4rf32w
  Fetch: 2.01s.  Build: 3m 32.73s.  Total: 3m 34.74s.
[+] /home/runderwood/git/contrib/spack/opt/spack/linux-fedora36-icelake/gcc-12.1.1/adios2-2.8.3-m7datpc2p3xb3ilgnmst2tpgiz4rf32w

```
